### PR TITLE
ESP32-C3 USB serial driver

### PR DESCRIPTION
**Pull request by martinjaeger**
This is a first attempt to get support for the built-in USB serial of the ESP32-C3. It is not a full-featured USB interface, that's why I put the drivers under `serial` instead of implementing them on top of the USB stack as CDC-ACM.

This is a single-instance driver, so we don't need the `const struct device *dev` in most of the functions.

Currently the driver does not compile because the header `hal/usb_serial_jtag_ll.h` is not existing in `hal_espressif`, but only in the original `esp-idf`.

@sylvioalves Could you have a look if the general approach of the driver looks fine or if there is a more suitable way to support this special serial interface? And which is the correct way to get the headers from ESP-IDF into `hal_espressif`?

### CODE REVIEW

**Comment by sylvioalves** [comment]
@martinjaeger, sure, I'll take a look. `hal_espressif` will be update to 4.4.1 soon, which will add missing dependencies.

**Comment by martinjaeger** [comment]
@sylvioalves Sounds good. I got it working with some dirty hacks (copying some files from ESP-IDF to locations where they don't belong). Just let me know when the `hal_espressif` update has been finished and I'll update the PR so that you can review a working version.

**Comment by lgl88911** [comment]
@sylvioalves After hal_espressif update to 4.4.1, Whether to support west flash via USB serial of esp32c3?

**Comment by martinjaeger** [comment]
@sylvioalves @glaubermaroto can you give any update on this?

If you don't have the time to update `hal_espressif` I'm fine with that and I'll come up with a workaround. But I'd really like to make some progress here.

@lgl88911 Yes, flashing via USB serial should work after this has been finished. I've successfully used it in my fork already.

**Comment by martinjaeger** [comment]
@lgl88911 FYI, here is an out-of-tree driver implementation which works very well (both for flashing and serial input/output): https://github.com/LibreSolar/bms-firmware/commit/1f1a53b33451bb4206a10fe00faeb92c69514605

**Comment by lgl88911** [comment]
> @lgl88911 FYI, here is an out-of-tree driver implementation which works very well (both for flashing and serial input/output): [LibreSolar/bms-firmware@1f1a53b](https://github.com/LibreSolar/bms-firmware/commit/1f1a53b33451bb4206a10fe00faeb92c69514605)

That's great, before the main branch is modified, it is painless to use USB serial. Also ask me if I need to upgrade hal_espressif and esp-idf to use it? In my understanding, esp32c3 needs specific esp-idf support to use USB serial

**Comment by martinjaeger** [comment]
> @martinjaeger, can you remove the "WIP:" from commit message and add missing sign-off id?

Sure. Was the HAL updated already without me noticing so that it should work now? If yes, I can clean up this PR and make it ready for review.

**Comment by sylvioalves** [comment]
Not yet, I'll bring those USB files before submitting the whole HAL update.

> 

PTAL https://github.com/zephyrproject-rtos/hal_espressif/pull/118

**Comment by martinjaeger** [comment]
> Not yet, I'll bring those USB files before submitting the whole HAL update.

Excellent, thanks. Please note also my other PR for the UART1 pinctrl. I'd like to backport that fix to v3.1-branch, so it would be nice if that could get merged quickly as well.

**Comment by sylvioalves** [comment]
@martinjaeger Is anything else needed to complete this?

**Comment by martinjaeger** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/46668 has to be merged.

**Review by sylvioalves — DISMISSED** [review]

**Review by carlescufi — DISMISSED** [review]
LGTM

**Comment by carlescufi** [comment]
@martinjaeger please rebase

**Comment by martinjaeger** [comment]
> @martinjaeger please rebase

Sorry, currently on vacation without laptop. If it's possible for someone else to rebase the PR, I'd appreciate that. Otherwise will do as soon as I'm back.

**Comment by fabiobaltieri** [comment]
> > @martinjaeger please rebase
> 
> Sorry, currently on vacation without laptop. If it's possible for someone else to rebase the PR, I'd appreciate that. Otherwise will do as soon as I'm back.

Hey I rebased for you, easy enough, trivial conflict in [esp32c3.dtsi](https://github.com/zephyrproject-rtos/zephyr/pull/44045/files#diff-4ef0770e2844cf23c1eb6976bfb64e1c127c4f664fcaa9b5babc4e4919866d2f)

**Review by fabiobaltieri — DISMISSED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by sylvioalves — APPROVED** [review]

File: drivers/serial/serial_esp32_usb.c [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+	.irq_err_disable = serial_esp32_usb_irq_err_disable,
+	.irq_is_pending = serial_esp32_usb_irq_is_pending,
+	.irq_update = serial_esp32_usb_irq_update,
+	.irq_callback_set = serial_esp32_usb_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+#define ESP32_UART_INIT(idx)                                                                       \
+	static const DRAM_ATTR struct serial_esp32_usb_config serial_esp32_usb_cfg_port_##idx = {  \
+		.uart_num = DT_INST_PROP(idx, peripheral),                                         \
```
Comment (gmarull) [inline drivers/serial/serial_esp32_usb.c:269]: I can't see this being used

File: drivers/serial/serial_esp32_usb.c [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+#include <device.h>
+#include <soc.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/interrupt_controller/intc_esp32c3.h>
+#include <zephyr/drivers/clock_control.h>
+#include <errno.h>
+#include <zephyr/sys/util.h>
+#include <esp_attr.h>
+
+#define ISR_HANDLER isr_handler_t
```
Comment (gmarull) [inline drivers/serial/serial_esp32_usb.c:22]: nit: why is this needed? can't we just do `(isr_handler_t)`?

File: drivers/serial/serial_esp32_usb.c [hunk 3]
```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2022 Libre Solar Technologies GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_usb_serial
+
+#include "stubs.h"
```
Comment (gmarull) [inline drivers/serial/serial_esp32_usb.c:9]: doesn't seem to be a local header, please use `<>`

**Review by gmarull — CHANGES_REQUESTED** [review]

**Comment by gmarull** [comment]
too late :-( please check my comments anyway and send a follow-up PR where applicable.

**Comment by martinjaeger** [comment]
Thanks @gmarull for your thorough review. There were indeed some left-overs from porting the "normal" UART driver to the special USB serial peripheral.

See follow-up PR with some additional clean-up: #47750
